homeassistant/components/notify/pushsafer.py

```diff
@@ -5,43 +5,171 @@
+    ATTR_TITLE, ATTR_TITLE_DEFAULT, ATTR_TARGET, ATTR_DATA,
+    PLATFORM_SCHEMA, BaseNotificationService)
 import homeassistant.helpers.config_validation as cv
-
 _LOGGER = logging.getLogger(__name__)
 _RESOURCE = 'https://www.pushsafer.com/api'
+_ALLOWED_IMAGES = ['image/gif', 'image/jpeg', 'image/png']
 
 CONF_DEVICE_KEY = 'private_key'
+CONF_TIMEOUT = 15
+
+# Top level attributes in 'data'
+ATTR_SOUND = 'sound'
+ATTR_VIBRATION = 'vibration'
+ATTR_ICON = 'icon'
+ATTR_ICONCOLOR = 'iconcolor'
+ATTR_URL = 'url'
+ATTR_URLTITLE = 'urltitle'
+ATTR_TIME2LIVE = 'time2live'
+ATTR_PICTURE1 = 'picture1'
 
-DEFAULT_TIMEOUT = 10
+ATTR_SOUND_DEFAULT = ''
+ATTR_VIBRATION_DEFAULT = ''
+ATTR_ICON_DEFAULT = ''
+ATTR_ICONCOLOR_DEFAULT = ''
+ATTR_URL_DEFAULT = ''
+ATTR_URLTITLE_DEFAULT = ''
+ATTR_TIME2LIVE_DEFAULT = ''
+ATTR_PICTURE1_DEFAULT = ''
+
+# Attributes contained in picture1
+ATTR_PICTURE1_URL = 'url'
+ATTR_PICTURE1_PATH = 'path'
+ATTR_PICTURE1_USERNAME = 'username'
+ATTR_PICTURE1_PASSWORD = 'password'
+ATTR_PICTURE1_AUTH = 'auth'
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Required(CONF_DEVICE_KEY): cv.string,
 })
-
-
 def get_service(hass, config, discovery_info=None):
     """Get the Pushsafer.com notification service."""
-    return PushsaferNotificationService(config.get(CONF_DEVICE_KEY))
-
-
+    return PushsaferNotificationService(config.get(CONF_DEVICE_KEY),
+                                        hass.config.is_allowed_path)
 class PushsaferNotificationService(BaseNotificationService):
     """Implementation of the notification service for Pushsafer.com."""
 
-    def __init__(self, private_key):
+    def __init__(self, private_key, is_allowed_path):
         """Initialize the service."""
         self._private_key = private_key
+        self.is_allowed_path = is_allowed_path
 
     def send_message(self, message='', **kwargs):
-        """Send a message to a user."""
+        """Send a message to a device."""
+        _LOGGER.info("Sending message")
+
+        """Send a message to specified target.
+        If no target specified (group or device),
+        a push will be sent to all devices
+        parameter d for devices is ignored here
+        """
+        if kwargs.get(ATTR_TARGET) is None:
+            targets = ["a"]
+            _LOGGER.debug("No target specified. Sending push to all")
+        else:
+            targets = kwargs.get(ATTR_TARGET)
+            _LOGGER.debug("%s target(s) specified", len(targets))
+
         title = kwargs.get(ATTR_TITLE, ATTR_TITLE_DEFAULT)
-        payload = {'k': self._private_key, 't': title, 'm': message}
-        response = requests.get(_RESOURCE, params=payload,
-                                timeout=DEFAULT_TIMEOUT)
-        if response.status_code != 200:
-            _LOGGER.error("Not possible to send notification")
+        data = kwargs.get(ATTR_DATA)
+
+        if data is None:
+            data = {}
+        """ Converting the specified image to base64"""
```

Comment:
Use `# Your comment` for comments. Only the first comment in a method is a triple quoted string and is used as pydoc.